terminal-view/src/main/java/com/termux/view/TerminalView.java

```diff
@@ -0,0 +1,1128 @@
+import android.text.InputType;
+import android.text.TextUtils;
+import android.util.AttributeSet;
+import android.view.ActionMode;
+import android.view.HapticFeedbackConstants;
+import android.view.InputDevice;
+import android.view.KeyCharacterMap;
+import android.view.KeyEvent;
+import android.view.MotionEvent;
+import android.view.View;
+import android.view.ViewConfiguration;
+import android.view.ViewTreeObserver;
+import android.view.accessibility.AccessibilityManager;
+import android.view.autofill.AutofillValue;
+import android.view.inputmethod.BaseInputConnection;
+import android.view.inputmethod.EditorInfo;
+import android.view.inputmethod.InputConnection;
+import android.widget.Scroller;
+
+import androidx.annotation.RequiresApi;
+
+import com.termux.terminal.KeyHandler;
+import com.termux.terminal.TerminalEmulator;
+import com.termux.terminal.TerminalSession;
+import com.termux.view.textselection.TextSelectionCursorController;
+
+/** View displaying and interacting with a {@link TerminalSession}. */
+public final class TerminalView extends View {
+
+    /** Log terminal view key and IME events. */
+    private static boolean TERMINAL_VIEW_KEY_LOGGING_ENABLED = false;
+
+    /** The currently displayed terminal session, whose emulator is {@link #mEmulator}. */
+    public TerminalSession mTermSession;
+    /** Our terminal emulator whose session is {@link #mTermSession}. */
+    public TerminalEmulator mEmulator;
+
+    public TerminalRenderer mRenderer;
+
+    public TerminalViewClient mClient;
+
+    private TextSelectionCursorController mTextSelectionCursorController;
+
+    private Handler mTerminalCursorBlinkerHandler;
+    private TerminalCursorBlinkerRunnable mTerminalCursorBlinkerRunnable;
+    private int mTerminalCursorBlinkerRate;
+    private boolean mCursorInvisibleIgnoreOnce;
+    public static final int TERMINAL_CURSOR_BLINK_RATE_MIN = 100;
+    public static final int TERMINAL_CURSOR_BLINK_RATE_MAX = 2000;
+
+    /** The top row of text to display. Ranges from -activeTranscriptRows to 0. */
+    int mTopRow;
+    int[] mDefaultSelectors = new int[]{-1,-1,-1,-1};
+
+    float mScaleFactor = 1.f;
+    final GestureAndScaleRecognizer mGestureRecognizer;
+
+    /** Keep track of where mouse touch event started which we report as mouse scroll. */
+    private int mMouseScrollStartX = -1, mMouseScrollStartY = -1;
+    /** Keep track of the time when a touch event leading to sending mouse scroll events started. */
+    private long mMouseStartDownTime = -1;
+
+    final Scroller mScroller;
+
+    /** What was left in from scrolling movement. */
+    float mScrollRemainder;
+
+    /** If non-zero, this is the last unicode code point received if that was a combining character. */
+    int mCombiningAccent;
+
+    private final boolean mAccessibilityEnabled;
+
+    private static final String LOG_TAG = "TerminalView";
+
+    public TerminalView(Context context, AttributeSet attributes) { // NO_UCD (unused code)
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 28. By comparison, 99% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test. We recommend that you simplify this method or break it into multiple methods.